blog/2022-05-11-SDL-historization/2022-05-011-historization.md

```diff
@@ -0,0 +1,340 @@
+      code = """
+        import org.apache.spark.sql.{DataFrame, SparkSession}
+        (session:SparkSession, options:Map[String, String], df:DataFrame, dataObjectId:String) => {
+          import session.implicits._
+          df.dropDuplicates(Seq("id"))
+        }
+      """
+   }]
+   metadata {
+      feed = download
+    }
+  }
+  #...
+}
+```
+
+The full configuration looks like [chess.conf](chess.conf). Note that there are already further dataObjects and actions defined, described and used later. 
+
+## Run
+
+The Pod with metastore, polynote and the "external" SQL server should already being running. Now the SDLB container is launched within the same POD and the action histData injects the data into the data lake:
+
+```Bash
+podman run --hostname localhost -e SPARK_LOCAL_HOSTNAME=localhost --rm --pod sdl_sql -v ${PWD}/data:/mnt/data -v ${PWD}/target:/mnt/lib -v ${PWD}/config:/mnt/config sdl-spark --config /mnt/config --feed-sel ids:histData
+```
+
+The data can be inspected using the Polynote notebook (`chessData.ipynb` downloaded above), which can be launched using [Polynote (click here)](http://localhost:8192/notebook/sql_data_monitor.ipynb). 
+
+Now, let's assume a change in the source database. Here the `victory_status` `outoftime` is renamed to `overtime`. Furthermore one entry is deleted. Run script using:
+
+```Bash
+podman exec -it mssql /opt/mssql-tools/bin/sqlcmd -S mssqlserver -U sa -P '%abcd1234%' -i /config/db_mod_chess.sql
+```
+Note: 1680 rows were changed.
+
+Furthermore, the data lake gets updated using the same command as above:
+```Bash
+podman run --hostname localhost -e SPARK_LOCAL_HOSTNAME=localhost --rm --pod sdl_sql -v ${PWD}/data:/mnt/data -v ${PWD}/target:/mnt/lib -v ${PWD}/config:/mnt/config sdl-spark --config /mnt/config --feed-sel ids:histData
+```
+
+In the [Polynote (click here)](http://localhost:8192/notebook/sql_data_monitor.ipynb) sql_data_monitor.ipynb, the data lake table can be inspected again. The The table and its additional columns are presented, as well as the original and updated rows for a modified row (*id = QQ3iIM2V*) and the deleted row (*id = 009mKOEz*).
+
+![polynote example output](historization_res.png)
+
+## Change Data Capture
+
+So far the whole table is collected and compared with the existing data lake table. 
+Various databases support Change Data Capture (CDC). Which already keeps track of changes similar to the above presented historization feature of SDL. 
+this can be utilized to only gathering database updates. Thus the amount of transferred and compared data can be significantly reduced. Therewith, only data changed (created, modified, or deleted) since the last syncronisation will be send. Therefore, a status is attached to the successful stream. This **state** is handled in SDLB. 
+
+In the following, an example is presented utilizing the MS SQL server CDC feature. Since the implemented JDBC connector cannot handle CDC data, an Airbyte connector is utilized. 
```

Comment:
```suggestion
In the following, an example is presented utilizing the MS SQL server CDC feature. Since the implemented JDBC connector cannot handle CDC data, an Airbyte connector is utilized, see [Airbyte CDC](https://docs.airbyte.com/understanding-airbyte/cdc/) for details.
```